core/shared/platform/include/platform_api_extension.h

```diff
@@ -1111,6 +1111,451 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
+ *
+ * @param handle the directory handle
+ * @param path the relative path of the directory to create
+ */
+__wasi_errno_t
+os_mkdirat(os_handle handle, const char *path);
+
+/**
+ * Rename a file or directory.
+ *
+ * @param old_handle the directory handle from which to resolve the old path
+ * @param old_path the source path to rename
+ * @param new_handle the directory handle from which to resolve the destination
+ * path
+ * @param new_path the destination path to which to rename the file or directory
+ */
+__wasi_errno_t
+os_renameat(os_handle old_handle, const char *old_path, os_handle new_handle,
+            const char *new_path);
+
+/**
+ * Unlink a file or directory.
+ *
+ * @param handle the directory handle from which to resolve the path
+ * @param path the relative path of the file or directory to unlink
+ * @param is_dir whether the provided handle refers to a directory or file
+ */
+__wasi_errno_t
+os_unlinkat(os_handle handle, const char *path, bool is_dir);
+
+/**
+ * Move the read/write offset of an open file.
+ *
+ * @param handle the associated file handle for which to adjust the offset
+ * @param offset the number of bytes to adjust the offset by
+ * @param whence the position whence to adjust the offset
+ * @param new_offset a pointer in which to store the new offset
+ */
+__wasi_errno_t
+os_lseek(os_handle handle, __wasi_filedelta_t offset, __wasi_whence_t whence,
+         __wasi_filesize_t *new_offset);
+
+/**
+ * Provide file advisory information for the given handle. This is similar to
+ * the POSIX function posix_fadvise.
+ *
+ * @param handle the associated file handle for which to provide advisory
+ * information
+ * @param offset the offset within the file to which the advisory
+ * information applies
+ * @param length the length of the region for which the advisory information
+ * applies
+ * @param advice the advice to provide
+ */
+__wasi_errno_t
+os_fadvise(os_handle handle, __wasi_filesize_t offset, __wasi_filesize_t length,
+           __wasi_advice_t advice);
+
+/**
+ * Determine if the given handle refers to a terminal device. __WASI_ESUCCESS
+ * will be returned if the handle is associated with a terminal device,
+ * otherwise an appropriate error code will be returned.
+ *
+ * @param handle
+ */
+__wasi_errno_t
+os_isatty(os_handle handle);
+
+/**
+ * Returns an internally used WASI libc handle for the given raw handle to
+ * STDIN. If the input handle is invalid (e.g. a file descriptor with a value
+ * of -1 on Linux), it will default to the platform-defined STDIN.
+ *
+ * @param stdio_handle a raw handle to STDIN
+ *
+ * @return a handle to STDERR
+ */
+os_handle
+os_get_stdin_handle(os_raw_handle stdio_handle);
+
+/**
+ * Returns an internally used WASI libc handle for the given raw handle to
+ * STDOUT. If the input handle is invalid (e.g. a file descriptor with a value
+ * of -1 on Linux), it will default to the platform-defined STDOUT.
```

Comment:
this interface mixes a few separate functionalities:
* "is invalid" check
* type conversion
* get stdin

i feel it's more straightforward to have separate apis.

Reply:
Yes, you're right. Unfortunately C doesn't have default parameters for functions 😄 which is really what I wanted in this function.

I'll split it up into two separate functions:
`os_convert_raw_handle` to convert a `os_raw_handle` into an `os_handle`
and `os_get_stdout_handle` etc. but without any arguments. Then we can explicitly check when initializing WASI whether the user-provided handle is valid and use the platform default stdio handles if it is.